File: tools/generate_changelog.php
<?php

define('REPOROOT', 'http://svn.abisource.com');
define('TAGROOT', REPOROOT . '/abiword/tags');

function help() {
	die('Usage: php ' . $_SERVER['argv'][0] . ' <x.y.z>' . "\n");
}

function get_tag_url($major, $minor, $micro) {
	return TAGROOT . "/release-$major.$minor.$micro";
}

function get_tag_copy_rev($tag_url) {
	return trim(`svn log --stop-on-copy --verbose --xml $tag_url | grep copyfrom-rev | cut -d "\"" -f 2`);
}

function get_changelog($from_rev, $to_rev, $url) {
	$xml = simplexml_load_string(`svn log -r$from_rev:$to_rev --xml --verbose $url`);
	if ($xml === FALSE)
		die("Error retrieving SVN log for range $from_rev:$to_rev from $url\n");

	return $xml;
}

function print_preamble($major, $minor, $micro) {
	print '<?
    include("../format.inc");
    printHeader("AbiWord v' . "$major.$minor.$micro" . ' ChangeLog");
?>

<h1>AbiWord v' . "$major.$minor.$micro" . ' Changelog</h1>
';
}

function print_postamble() {
	print '<?
    printFooter();
?>';
}

function is_backport_blocking($msg) {
	// check if this is an svnmerge.py blocked merge
	return preg_match('/^Blocked revision.*\.\.\.\.\.\.\.\.(.*)\.\.\.\.\.\.\.\..*$/ms', $msg);
}

function detect_embedded_commit($msg) {
	// check if this is an svnmerge.py recorded merge
	if (!preg_match('/^Recorded merge of.*\.\.\.\.\.\.\.\.(.*)\.\.\.\.\.\.\.\..*$/ms', $msg, $matches))
		return FALSE;

	$lines = explode("\n", trim($matches[1]));

	if (!preg_match('/.*\|(.*)\|.*\|.*$/', $lines[0], $matches))
		return FALSE;

	$author = trim($matches[1]);
	$msg = trim(implode(' ', array_slice($lines, 1)));

	return array($author, $msg);
}

function human_readable_author($author) {
	$table = array(
		'asfrent' => 'Andrei Sfrent',
		'cjl' => 'Chris Leonard',
		'dom' => 'Dominic Lachowicz',
		'hub' => 'Hubert Figui&egrave;re',
		'ib' => 'Ingo Br&uuml;ckl',
		'jbrefort' => 'Jean Br&eacute;fort',
		'monkeyiq' => 'Ben Martin',
		'msevior' => 'Martin Sevior',
		'pradeeban' => 'Kathiravelu Pradeeban',
		'slaroche' => 'Simon Larochelle',
		'strbafridrich' => 'Fridrich Strba',
		'uwog' => 'Marc Maurer',
		'volodymyr' => 'Volodymyr Rudyj',
		);
	if (array_key_exists($author, $table))
		return $table[$author];
	return $author;
}

function strip_backports($msg) {
	if (preg_match('/(.*)\(?(Please backport|Can be backported|Feel free to backport)\.?\)?(.*)/i', $msg, $matches))
		$msg = $matches[1] . $matches[3];

	return $msg;
}

function add_bugzilla_links($msg) {
	if (preg_match('/(.*?)((Bug\s|Fix\s)?(\d\d\d*))(.*)/', $msg, $matches)) {
		$msg = $matches[1] . '<a href="http://bugzilla.abisource.com/show_bug.cgi?id=' . $matches[4] . '">' . $matches[2] . '</a>' . $matches[5];
	}
	return $msg;
}

function classify_topic($paths, $msg) {
	$topic = 'Core';
	$sub_topic = '-';

	// path based classification
	foreach ($paths as $path) {
		// Feel free to expand this list to reduce the manual work

		if (preg_match('/.*\/po\/.*\.po/', $path)) {
			$topic = 'Internationalization';
			break;
		}

		if (preg_match('/.*\/opendocument\/.*/', $path)) {
			$topic = 'Import/Export';
			$sub_topic = 'OpenDocument';
			break;
		}

		if (preg_match('/.*\/openxml\/.*/', $path)) {
			$topic = 'Import/Export';
			$sub_topic = 'Office Open XML';
			break;
		}

		if (preg_match('/.*\/goffice\/.*/', $path)) {
			$topic = 'Plugins';
			$sub_topic = 'GNOME Office';
			break;
		}

		if (preg_match('/.*\/collab\/.*/', $path)) {
			$topic = 'Plugins';
			$sub_topic = 'Collaboration';
			break;
		}

		if (preg_match('/.*\/cocoa\/.*/', $path)) {
			$topic = 'Interface';
			$sub_topic = 'Mac OSX';
			break;
		}

		if (preg_match('/.*MSWrite.*/', $path)) {
			$topic = 'Plugins';
			$sub_topic = 'MS Write';
			break;
		}
	}

	// commit message based classification
	if (preg_match('/.*translation.*/', $msg)) {
		$topic = 'Internationalization';
	}

	if (preg_match('/.*buildfix.*/i', $msg)) {
		$topic = 'Development';
	}

	if (preg_match('/.*bump version.*/i', $msg)) {
		$topic = 'Development';
	}

	return array($topic, $sub_topic);
}

function parse_cherry_pick($msg) {
	$re = '/(Backported|Cherry[- ]pick(ed)?) from trunk@[0-9]*[.]*[0-9]*\.?( [Bb]y ([a-z]*)\.?)?/';
	if (preg_match($re, $msg, $matches) == 1) {
		$msg = preg_replace($re, '', $msg);
		if (count($matches) == 5) {
			return array($msg, $matches[4]);
		}
		return array($msg);
	}
	return FALSE;
}

function parse_logentries($xml) {
	$result = array();
	foreach ($xml->logentry as $logentry) {
		// get the author
		$author = trim((string)$logentry->author);

		// get the commit message if this commit was not a blocking of
		// blocking (and strip out all the svnmerge.py output if present)
		$msg = trim((string)$logentry->msg);
		if (is_backport_blocking($msg))
			continue;
		$commit = detect_embedded_commit($msg);
		if ($commit !== FALSE) {
			list($author, $msg) = $commit;
		}

		// get the affected paths
		$paths = array();
		foreach ($logentry->paths->path as $path) {
			$paths[] = (string)$path;
		}

	        $cherrypick = parse_cherry_pick($msg);
		if ($cherrypick != FALSE) {
			$msg = $cherrypick[0];
			if (count($cherrypick) > 1) {
				$author = $cherrypick[1];
			}
		}

		// convert the svn usernames into real-life names
		$author = human_readable_author($author);

		// strip double spaces from commit messages
		$msg = preg_replace('/\s\s+/', ' ', $msg);

		// strip the common "Please backport" and friends from commit messages
		$msg = strip_backports($msg);

		// add convenient links to bugzilla for bug numbers
		$msg = add_bugzilla_links($msg);

		// figure out which topic this commit belongs to
		list($topic, $sub_topic) = classify_topic($paths, $msg);

		$result[$topic][$sub_topic][] = array(trim($msg), $author);
	}
	return $result;
}

function print_changelog_section($logentries) {
	print "<ul>\n";
	foreach ($logentries as $logentry) {
		list($msg, $author) = $logentry;
		print "  <li>$msg ($author)</li>\n";
	}
	print "</ul>\n";
}

function print_changelog($xml) {
	$topics = parse_logentries($xml);
	foreach ($topics as $topic => $sub_topics) {
		print "<h2>$topic</h2>\n";

		// print items without a subtopic first...
		if (isset($sub_topics['-']))
			print_changelog_section($sub_topics['-']);

		// ... and then print all the subtopics
		foreach (array_diff_key($sub_topics, array('-' => '')) as $sub_topic => $logentries) {
			print "<h3>$sub_topic</h3>\n";
			print_changelog_section($logentries);
		}
	}
}

function main() {
	if (sizeof($_SERVER['argv']) != 2)
		help();

	// get the version number
	list($major, $minor, $micro) = explode('.', $_SERVER['argv'][1]);
	if (!is_numeric($major) || !is_numeric($minor) || !is_numeric($micro))
		help();

	if ($micro == 0)
		die("Sorry, you'll have to do x.y.0 releases by hand :)\n");


	// get the revision range between the current and previous tag
	$tag_url = get_tag_url($major, $minor, $micro);
	$prev_tag_url = get_tag_url($major, $minor, $micro - 1);
	
	$tag_copy_rev = get_tag_copy_rev($tag_url);
	$prev_tag_copy_rev = get_tag_copy_rev($prev_tag_url) + 1;

	// retrieve the changelog for this release
	$xml = get_changelog($prev_tag_copy_rev, $tag_copy_rev, $tag_url);

	// process the changelog and output a nicely formatted piece of HTML
	print_preamble($major, $minor, $micro);
	print_changelog($xml);
	print_postamble();
}

main();

?>

File: tools/generate_changelog.py
#!/bin/env python3

import os
import re
import sys
import git


def get_start_rev(major, minor, micro):
    if int(micro) == 0:
        raise Exception("Wrong version {}.{}.{}".format(major, minor, micro))
    return "release-{}.{}.{}".format(major, minor, int(micro) - 1)

def get_end_rev(major, minor, micro):
    return "release-{}.{}.{}".format(major, minor, micro)

# Remove the "cherry-picked from" lines
def strip_cherry_pick(msg):
    regexp = re.compile(r"^\(?cherry[ \-]pick(ed)? from.*$", flags=re.I|re.M)
    msg = regexp.sub("", msg)
    return msg

# Strip the backport mentions
def strip_backports(msg):
    regexp = re.compile(r"\(?(Please backport|Can be backported|Feel free to backport)\.?\)?", flags=re.I)
    msg = regexp.sub("", msg)
    return msg

# Strip the svn ids. Shouldn't be needed past 3.0.3
def strip_svn_id(msg):
    regexp = re.compile(r"^git-svn-id:.*$", flags=re.M)
    msg = regexp.sub("", msg)
    return msg

# Strip signoff mentions
def strip_signoff(msg):
    regexp = re.compile(r"^Signed-off-by:.*$", flags=re.M)
    msg = regexp.sub("", msg)
    return msg

# Linkify bugzilla links
def add_bugzilla_links(msg):
    m = re.match(r"(.*?)((Bug\s|Fix\s)?(\d\d\d+))(.*)", msg)
    if m:
        msg = m.group(1) + "<a href=\"http://bugzilla.abisource.com/show_bug.cgi?id=" + m.group(4) + "\">" + m.group(2) + "</a>" + m.group(5)

    return msg


def classify_topic(paths, msg):
    topic = "Core"
    sub_topic = "-"

    for path in paths:
        if re.match(r".*\/opendocument\/.*", path):
            topic = 'Import/Export'
            sub_topic = 'OpenDocument'
            break

        if re.match(r".*\/openxml\/.*", path):
            topic = 'Import/Export'
            sub_topic = 'Office Open XML'
            break

        if re.match(r".*\/goffice\/.*", path):
            topic = 'Plugins'
            sub_topic = 'GNOME Office'
            break

        if re.match(".*\/collab\/.*", path):
            topic = 'Plugins';
            sub_topic = 'Collaboration';
            break;

        if re.match(r".*\/cocoa\/.*", path):
            topic = 'Interface';
            sub_topic = 'Mac OSX';
            break;

        if re.match(".*MSWrite.*", path):
            topic = 'Plugins';
            sub_topic = 'MS Write';
            break;

    # commit message based classification
    if re.match(r".*translation.*", msg):
        topic = 'Internationalization'

    regexp = re.compile(r".*buildfix.*", "i")
    if regexp.match(msg):
        topic = "Development"

    regexp = re.compile(r".*bump version.*", "i")
    if regexp.match(msg):
        topic = 'Development'

    return (topic, sub_topic)


def print_preamble(major, minor, micro):
    print("<?")
    print("include(\"../format.inc\");")
    print("printHeader(\"AbiWord v{}.{}.{} ChangeLog\");".format(major, minor, micro))
    print("?>")
    print("<h1>AbiWord v{}.{}.{} Changelog</h1>".format(major, minor, micro))

def print_postamble():
    print("<?\nprintFooter();\n?>")

# get_commits from repository from_rev..to_rev
# return the list of commits
def get_commits(repo, from_rev, to_rev):
    # if we can't resolve to_rev we'll consider the tip to the branch
    try:
        t = repo.commit(to_rev)
    except:
        to_rev = ""

    try:
        commits = list(repo.iter_commits(rev="{}..{}".format(from_rev, to_rev)))
    except:
        commits = []
    return commits

# Print the changelog for the commits list
def print_changelog(repo, commits):
    print("<ul>")
    for commit in commits :
        c = repo.commit(commit)
        msg = strip_svn_id(c.message)
        msg = strip_backports(msg)
        msg = strip_cherry_pick(msg)
        msg = strip_signoff(msg)
        msg = add_bugzilla_links(msg)
        msg = msg.strip()
        print("  <li>{} ({})</li>".format(msg, c.author))
    print("</ul>")


try:
    # get the version from argument one
    if len(sys.argv) < 2:
        raise Exception("Please pass the version you want to generate the changelog for")

    argv1 = sys.argv[1]
    (major, minor, micro) = argv1.split(".", maxsplit=3);

    start = get_start_rev(major, minor, micro)
    end = get_end_rev(major, minor, micro)

    repo = git.Repo.init(os.path.join("."))

    commits = get_commits(repo, start, end)
    if len(commits) == 0:
        raise Exception("No commit founds.")

    print_preamble(major, minor, micro)
    print_changelog(repo, commits)
    print_postamble()

except:
    raise
